phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexBuildTimestampIT.java

```diff
@@ -0,0 +1,236 @@
+        props.setProperty("CurrentSCN", Long.toString(clock1.initialTime()));
+        conn = DriverManager.getConnection(getUrl(), props);
+        ResultSet rs = conn.createStatement().executeQuery("select * from " + tableName);
+        assertFalse(rs.next());
+        conn.close();
+
+        props.setProperty("CurrentSCN", Long.toString(clock2.initialTime()));
+        conn = DriverManager.getConnection(getUrl(), props);
+        rs = conn.createStatement().executeQuery("select * from " + tableName);
+
+        assertTrue(rs.next());
+        assertEquals("aaa", rs.getString(1));
+        assertEquals("abc", rs.getString(2));
+        assertNotNull(rs.getDate(3));
+
+        assertFalse(rs.next());
+        conn.close();
+
+        props.setProperty("CurrentSCN", Long.toString(clock2.currentTime()));
+        conn = DriverManager.getConnection(getUrl(), props);
+        rs = conn.createStatement().executeQuery("select * from " + tableName);
+
+        assertTrue(rs.next());
+        assertEquals("aaa", rs.getString(1));
+        assertEquals("abc", rs.getString(2));
+        assertNotNull(rs.getDate(3));
+
+        assertTrue(rs.next());
+        assertEquals("bbb", rs.getString(1));
+        assertEquals("bcd", rs.getString(2));
+        assertNotNull(rs.getDate(3));
+        assertFalse(rs.next());
+        conn.close();
+    }
+
+    @Test
+    public void testCellTimestamp() throws Exception {
+        EnvironmentEdgeManager.reset();
+        MyClock clock1 = new MyClock(100000);
+        MyClock clock2 = new MyClock(200000);
+        String dataTableName = generateUniqueName();
+        populateTable(dataTableName, clock1, clock2);
+
+        MyClock clock3 = new MyClock(300000);
+        EnvironmentEdgeManager.injectEdge(clock3);
+
+        Properties props = new Properties();
+        props.setProperty(QueryServices.ENABLE_SERVER_SIDE_MUTATIONS, "true");
+        Connection conn = DriverManager.getConnection(getUrl(), props);
+
+        String indexName = generateUniqueName();
+        conn.createStatement().execute("CREATE "+ (localIndex ? "LOCAL " : "") + " INDEX " + indexName + " on " +
+                dataTableName + " (val) include (ts)" + (async ? "ASYNC" : ""));
+
+        conn.close();
+
+        if (async) {
+            // run the index MR job.
+            IndexToolIT.runIndexTool(true, false, null, dataTableName, indexName);
+        }
+
+        // Verify the index timestamps via Phoenix
+        String selectSql = String.format("SELECT * FROM %s WHERE val = 'abc'", dataTableName);
+        conn = DriverManager.getConnection(getUrl());
+        // assert we are pulling from index table
+        assertExplainPlan(conn, localIndex, selectSql, dataTableName, indexName);
+
+        ResultSet rs = conn.createStatement().executeQuery(selectSql);
+        assertTrue (rs.next());
+        assertTrue(rs.unwrap(PhoenixResultSet.class).getCurrentRow().getValue(0).getTimestamp() >= clock1.initialTime());
```

Comment:
@kadirozde - maybe I'm misunderstanding the test, but shouldn't this be checking against an _upper_ bound to make sure that the index was populated with the initial timestamp of 'abc' (and in the second check 'bcd')?

Reply:
Thx for noticing it. I will correct it.